modules/phase_field/src/auxkernels/EBSDReaderAvgDataAux.C

```diff
@@ -5,38 +5,42 @@
 /*             See LICENSE for full restrictions                */
 /****************************************************************/
 #include "EBSDReaderAvgDataAux.h"
+#include "EBSDReader.h"
+#include "GrainTrackerInterface.h"
 
 template<>
 InputParameters validParams<EBSDReaderAvgDataAux>()
 {
   InputParameters params = validParams<AuxKernel>();
   params.addRequiredParam<UserObjectName>("ebsd_reader", "The EBSDReader GeneralUserObject");
+  params.addRequiredParam<UserObjectName>("grain_tracker", "The GrainTracker UserObject");
   MooseEnum field_types = EBSDAccessFunctors::getAvgDataFieldType();
   params.addRequiredParam<MooseEnum>("data_name", field_types, "The averaged data to be extracted from the EBSD data by this AuxKernel");
+  params.addParam<Real>("invalid", -1.0, "Value to return for points without active grains.");
   return params;
 }
 
 EBSDReaderAvgDataAux::EBSDReaderAvgDataAux(const InputParameters & parameters) :
     AuxKernel(parameters),
     _ebsd_reader(getUserObject<EBSDReader>("ebsd_reader")),
+    _grain_tracker(getUserObject<GrainTrackerInterface>("grain_tracker")),
     _data_name(getParam<MooseEnum>("data_name")),
-    _val(_ebsd_reader.getAvgDataAccessFunctor(_data_name))
+    _val(_ebsd_reader.getAvgDataAccessFunctor(_data_name)),
+    _invalid(getParam<Real>("invalid"))
 {
 }
 
 Real
 EBSDReaderAvgDataAux::computeValue()
 {
-  // EBSD data is defined at element centroids, so this only makes
-  // sense as an Element AuxKernel
-  Point p = _current_elem->centroid();
+  // get the dominant grain for the current element/node
+  const int grain_id = _grain_tracker.getEntityValue(isNodal() ? _current_node->id() : _current_elem->id(),
+                                                     FeatureFloodCount::FieldType::UNIQUE_REGION, 0);
 
-  // get the EBSD Point data for the current element
-  const EBSDPointData & d = _ebsd_reader.getData(p);
+  // no grain found
+  if (grain_id < 0)
+    return _invalid;
```

Comment:
And what is `getEntityValue` supposed to return? I think this is the right thing to do (and it is basically the same thing you did in your current PR). I'm asking for the current dominant grain and I'm handling the case where no grain is found separately.